clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+import org.apache.kafka.common.message.StreamsGroupHeartbeatRequestData;
+import org.apache.kafka.common.message.StreamsGroupHeartbeatResponseData;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.StreamsGroupHeartbeatRequest;
+import org.apache.kafka.common.requests.StreamsGroupHeartbeatResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+
+public class StreamsGroupHeartbeatRequestManager implements RequestManager {
+
+    static class HeartbeatState {
+
+        private final StreamsMembershipManager membershipManager;
+        private final int rebalanceTimeoutMs;
+
+        private final StreamsRebalanceData streamsRebalanceData;
+
+        public HeartbeatState(final StreamsRebalanceData streamsRebalanceData,
+                              final StreamsMembershipManager membershipManager,
+                              final int rebalanceTimeoutMs) {
+
+            this.membershipManager = membershipManager;
+            this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+            this.streamsRebalanceData = streamsRebalanceData;
+        }
+
+        public StreamsGroupHeartbeatRequestData buildRequestData() {
+            StreamsGroupHeartbeatRequestData data = new StreamsGroupHeartbeatRequestData();
+            data.setGroupId(membershipManager.groupId());
+            data.setMemberId(membershipManager.memberId());
+            data.setMemberEpoch(membershipManager.memberEpoch());
+            membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+            StreamsGroupHeartbeatRequestData.Topology topology = new StreamsGroupHeartbeatRequestData.Topology();
+            topology.setSubtopologies(getTopologyFromStreams());
+            topology.setEpoch(streamsRebalanceData.topologyEpoch());
+            data.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+            data.setTopology(topology);
+            data.setProcessId(streamsRebalanceData.processId().toString());
+            data.setActiveTasks(Collections.emptyList());
+            data.setStandbyTasks(Collections.emptyList());
+            data.setWarmupTasks(Collections.emptyList());
+            streamsRebalanceData.endpoint().ifPresent(userEndpoint -> {
+                data.setUserEndpoint(new StreamsGroupHeartbeatRequestData.Endpoint()
+                    .setHost(userEndpoint.host())
+                    .setPort(userEndpoint.port())
+                );
+            });
+            data.setClientTags(streamsRebalanceData.clientTags().entrySet().stream()
```

Comment:
It would probably be nice to cache the results of transformations on these immutable fields, so that we don't have to redo it on every heartbeat. We could create this in private field in the constructor. Just a minor optimization, not sure if strictly necessary

Reply:
Actually, the client tags are only sent with the first full request and they do not change afterwards. However, this is only true because the client tags are set in Streams trough a config which is constant. The KIP says that the client tags are not sent if they do not change. So I was thinking to implement the request manager in such a way that it follows closer the KIP and does not assume the client tags are constant. This is also true for process ID and endpoint.
In any case I will consider this in the next PR since this PR is really just about sending the full request.